src/victory-util/transitions.js

```diff
@@ -0,0 +1,296 @@
+  const nextDataKeyed = getKeyedData(nextData);
+
+  return {
+    entering: getKeyedDataDifference(nextDataKeyed, oldDataKeyed),
+    exiting: getKeyedDataDifference(oldDataKeyed, nextDataKeyed)
+  };
+}
+
+/**
+ * If a parent component has animation enabled, calculate the transitions
+ * for any data of any child component that supports data transitions
+ * Data transitions are defined as any two datasets where data nodes exist
+ * in the first set and not the second, in the second and not the first,
+ * or both.
+ *
+ * @param  {Children}  oldChildren   this.props.children from old props
+ * @param  {Children}  nextChildren  this.props.children from next props
+ *
+ * @return {Object}                  Object with the following properties:
+ *                                    - nodesWillExit
+ *                                    - nodesWillEnter
+ *                                    - childrenTransitions
+ *                                    - nodesShouldEnter
+ */
+export function getInitialTransitionState(oldChildren, nextChildren) {
+  let nodesWillExit = false;
+  let nodesWillEnter = false;
+
+  // Children may be a single item, rather than an array.
+  oldChildren = [].concat(oldChildren);
+  nextChildren = [].concat(nextChildren);
+
+  const childrenTransitions = oldChildren.map((child, idx) => {
+    // TODO: Determine if/how we want to support variable-length children.
+    const nextChild = nextChildren[idx];
+    if (!nextChild || child.type !== nextChild.type) {
+      return {};
+    }
+
+    const { entering, exiting } =
+      child.type.supportsTransitions &&
+      getNodeTransitions(child.props.data, nextChild.props.data) ||
+      {};
+
+    nodesWillExit = nodesWillExit || !!exiting;
+    nodesWillEnter = nodesWillEnter || !!entering;
+
+    return { entering, exiting };
+  });
+
+  return {
+    nodesWillExit,
+    nodesWillEnter,
+    childrenTransitions,
+    // TODO: This may need to be refactored for the following situation.
+    //       The component receives new props, and the data provided
+    //       is a perfect match for the previous data and domain except
+    //       for new nodes. In this case, we wouldn't want a delay before
+    //       the new nodes appear.
+    nodesShouldEnter: false
+  };
+}
+
+
+function getInitialChildProps(animate, data) {
+  return {
+    data: data.map((datum) => assign({}, datum, animate.onExit.before(datum)))
+  };
+}
+
+function getChildPropsOnExit(animate, data, exitingNodes, cb) { // eslint-disable-line max-params
+  // Whether or not _this_ child has exiting nodes, we want the exit-
+  // transition for all children to have the same duration, delay, etc.
+  animate = assign({}, animate, animate.onExit);
+
+  if (exitingNodes) {
+    // After the exit transition occurs, trigger the animations for
+    // nodes that are neither exiting or entering.
+    animate.onEnd = cb;
+
+    // If nodes need to exit, all we want to do is to fade them out.
```

Comment:
The exit transition should be factored out into a function so we can generalize it later and expose it for user control. I could imagine a user wanting the bar exit transition to be `Object.assign({}, datum, { y: 0 })` for example.